src/zenml/stack/stack.py

```diff
@@ -489,11 +488,106 @@ def validate(self) -> None:
             remote execution) specified by the orchestrator of the stack
         - the `StackValidator` of each stack component has to validate the
             stack to make sure all the components are compatible with each other
+        - the stack must either have a properly associated artifact/metadata
+            store pair or reset the association.
+
+        Args:
+            reset_association: Flag to reset the previous associations between
+                an artifact store and a metadata store
+
+        Raises:
+            StackValidationError: If the artifact store and the metadata store
+                are not properly associated.
         """
+        from zenml.cli.utils import warning
+        from zenml.repository import Repository
+
         for component in self.components.values():
             if component.validator:
                 component.validator.validate(stack=self)
 
+        r = Repository()
+        a_associations = r.zen_store.get_store_associations_for_artifact_store(
+            self.artifact_store.uuid
+        )
+        if a_associations:
+            for a in a_associations:
+                if a.metadata_store_uuid != self.metadata_store.uuid:
+                    if reset_association:
+                        warning(
+                            f"Removing the association between given artifact "
+                            f"store {self.artifact_store.name} (uuid: "
+                            f"{self.artifact_store.uuid}) and the metadata "
+                            f"store (uuid: {a.metadata_store_uuid})."
+                        )
+                        r.zen_store.delete_store_association_for_artifact_and_metadata_store(
+                            artifact_store_uuid=self.artifact_store.uuid,
+                            metadata_store_uuid=a.metadata_store_uuid,
+                        )
+                    else:
+                        raise StackValidationError(
+                            f"The artifact store instance in your stack "
+                            f"'{self.artifact_store.name}' (uuid: "
+                            f"{self.artifact_store.uuid}) has been previously "
+                            f"associated with a different metadata store "
+                            f"(uuid: {a.metadata_store_uuid} in a different "
+                            f"stack. If either one of these stores are "
+                            f"previously populated, this might lead to various "
+                            f"problems. In order to solve this issue, you can "
+                            f"either create and use another artifact store "
+                            f"instance or use the '-r' flag when you "
+                            f"register/update a stack to reset the "
```

Comment:
Yes. I am still a bit unsure how this would affect the overall interaction with our pipelines however I do feel that we need to enforce some form of rules here. Both pairs of `populated AS/empty MS` and `empty AS/populated MS` lead to breaking problems. The flag is there for users who manually interact with these stores (like anyone who would empty the artifact stores themselves) to bypass this rule. While doing that, they have to break their old couplings.